lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -4629,8 +4629,33 @@ ConstraintResult GenericSignatureBuilder::addTypeRequirement(
                         ->getDependentType(getGenericParams());
 
       Impl->HadAnyError = true;
+      
+      bool shouldOfferFixit = false;
+      auto DMT = subjectType->getAs<DependentMemberType>();
+      auto GTPT = subjectType->getAs<GenericTypeParamType>();
+
+      if (GTPT) {
+        shouldOfferFixit = isa<ExtensionDecl>(GTPT->getDecl()
+                                                  ->getDeclContext());
+      }
+
+      if (DMT) {
+        shouldOfferFixit = isa<ExtensionDecl>(DMT->getRootGenericParam()
+                                                 ->getDecl()
+                                                 ->getDeclContext());
+      }
+      
       Diags.diagnose(source.getLoc(), diag::requires_conformance_nonprotocol,
                      subjectType, constraintType);
+      
+      if (shouldOfferFixit) {
+        auto subjectTypeName = DMT ? DMT->getName().str().str() :
```

Comment:
Is this going to be sufficiently qualified? What if it's `Assoc.Foo.Bar`?

Reply:
@jrose-apple Seems to work for the cases I tried. Do you have an example you think that could fail to work here or lead to an incorrect name?

Reply:
Can you add some tests to cover a Foo.Bar.Baz example just in case?

Reply:
@slavapestov Do you mean something like

```swift
class Foo {
    struct Bar<Baz> {}
}

extension Foo.Bar where Baz: String {}
```

?

Reply:
No, we're talking about

```swift
protocol HasFoo {
  associatedtype Foo
}
protocol HasAssoc {
  associatedtype Assoc: HasFoo
}

struct X<T: HasAssoc> {}
extension X where T.Assoc.Foo: String {}
```

Reply:
Validation testing has found a few cases where I think the fix-it won't work:

```swift
struct Foo<A, B> {
  let a: A, b : B
}

// note: use 'T == Int' to require 'T' to be 'Int'
typealias Alias<T: Int> = Foo<String, T>
```

I guess there needs to be some extra checks to handle cases like the one above.

Reply:
Heh. Yeah, completely *removing* a generic parameter is generally beyond the scope of this fix. I guess that's why it was constrained to extensions to begin with?

Reply:
@jrose-apple Thanks, that example leads to an incorrect diagnostic message (the note says `use 'Foo == String' to require 'Foo' to be 'String'` rather than `use 'T.Assoc.Foo == String' to require 'T.Assoc.Foo' to be 'String'`. I guess I could simply match the note with the error diagnostic - but it will sometimes say "Self.Bar" instead of just "Bar"., or just remove the "Self" prefix (if present) is another option (which I prefer).

I kept it to extensions for simplicity, I hadn't thought of this case, but I can either handle it or just do the simple fix now and build upon it in another PR.

Reply:
@jrose-apple Right, I guess I'll revert back to what I was doing before!

Reply:
@jrose-apple @slavapestov could you take a look again? I've added the above example in the test as well.